### PR TITLE
[oss] Upgrade GORM to latest major version.

### PR DESCRIPTION
**Pull request by vadimberezniker**
Summary of the changes:
  - Replace gorm.IsRecordNotFoundError(err) with errors.Is(err,
gorm.ErrRecordNotFound) 
  - Metrics are now exported by taking advantage of the GORM Callbacks
API that lets us hook into the GORM execution pipeline.
  - Add SQL template label to SQL error counter
  - Filter out "record not found" errors from the default logger

Part of addressing buildbuddy-io/buildbuddy-internal#250

**Version bump**: Minor

### CODE REVIEW

File: server/util/db/db.go [hunk 1]
```diff
@@ -140,88 +210,60 @@ func parseDatasource(datasource string) (string, string, error) {
 	return "", "", fmt.Errorf("No database configured -- please specify at least one in the config")
 }
 
+// sqlLogger is a GORM logger wrapper that supresses "record not found" errors.
 type sqlLogger struct {
-	logQueries bool
-	delegate   gorm.Logger
+	logLevel logger.LogLevel
+	delegate logger.Interface
 }
```
Comment (tylerwilliams) [inline server/util/db/db.go:217]: Another option here that might save you some lines is making delegate an anonymous field: (http://golangtutorials.blogspot.com/2011/06/anonymous-fields-in-structs-like-object.html)

I think this would let you remove the Info, Warn, and Error methods you've had to define that just call the delegate. You could still of course keep logLevel as a named member.

File: server/util/db/db.go [hunk 2]
```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
...
+	gdb.Callback().Query().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Raw().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Row().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Update().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+}
+
+func openDB(configurator *config.Configurator, dialect string, connString string) (*gorm.DB, error) {
+	var dialector gorm.Dialector
+	switch dialect {
+	case "sqlite3":
```
Comment (tylerwilliams) [inline server/util/db/db.go:170]: Are there consts we could use from gorm or that we could define ourselves for these? {"sqlite3", "mysql", "postgres"}?

File: server/util/db/db.go [hunk 3]
```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
...
+		if opStartTime, ok := v.(time.Time); ok {
+			metrics.SQLQueryDurationUsec.With(labels).Observe(float64(time.Now().Sub(opStartTime).Microseconds()))
+		}
+		// Ignore "record not found" errors as they don't generally indicate a
+		// problem with the server.
+		if db.Error != nil && !errors.Is(db.Error, gorm.ErrRecordNotFound) {
+			metrics.SQLErrorCount.With(labels).Inc()
+		}
+	}
+	gdb.Callback().Create().After("*").Register("buildbuddy:record_metrics", recordMetrics)
```
Comment (tylerwilliams) [inline server/util/db/db.go:159]: could we make the string "buildbuddy:record_metrics" a const?

File: server/util/db/db.go [hunk 4]
```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
...
+
+	// Add callback that runs after other callbacks that records executed queries and their durations.
+	recordMetrics := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		labels := prometheus.Labels{
+			metrics.SQLQueryTemplateLabel: db.Statement.SQL.String(),
+		}
+		v, _ := db.Statement.Settings.LoadAndDelete(gormStmtStartTimeKey)
```
Comment (tylerwilliams) [inline server/util/db/db.go:149]: If this code ignores the error returned by LoadAndDelete -- can you add a short comment about why this is OK? Just is sort of a code-smell to ignore errors like this so when we do it I think we should be explicit about why!

(seems like if there is an err, v is bad, then the query duration won't be logged which is fine! and we still will count the sqlerrorcount, which is what we want)

File: server/util/db/db.go [hunk 5]
```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
...
+
+	var l logger.Interface
+	l = sqlLogger{delegate: logger.Default, logLevel: logger.Warn}
+	if configurator.GetDatabaseConfig().LogQueries {
+		l = l.LogMode(logger.Info)
+	}
+	config := gorm.Config{
+		Logger: l,
+	}
+	gdb, err := gorm.Open(dialector, &config)
```
Comment (tylerwilliams) [inline server/util/db/db.go:191]: "l" is nil if LogQueries is not defined. That might be OK in this case -- but you could you rephrase this in a way that is maybe safer by doing:

config := gorm.Config{}
if configurator.GetDatabaseConfig().LogQueries {
                 l := sqlLogger{delegate: logger.Default, logLevel: logger.Warn}
		l = l.LogMode(logger.Info)
                config.Logger = l
}

gdb, err := gorm.Open(dialector, &config)

File: server/util/db/db.go [hunk 6]
```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
...
+func instrumentGORM(gdb *gorm.DB) {
+	// Add callback that runs before other callbacks that records when the operation began.
+	// We use this to calculate how long a query takes to run.
+	recordStartTime := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		db.Statement.Settings.Store(gormStmtStartTimeKey, time.Now())
+	}
+	gdb.Callback().Create().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
```
Comment (tylerwilliams) [inline server/util/db/db.go:134]: could "buildbuddy:record_op_start_time" be a const?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
Few nits but this LG overall!

File: server/util/db/db.go (hunk 1 above)
Comment (vadimberezniker) [reply server/util/db/db.go:217]: Good idea, thanks for the pointer

File: server/util/db/db.go (hunk 2 above)
Comment (vadimberezniker) [reply server/util/db/db.go]: Done.

File: server/util/db/db.go (hunk 4 above)
Comment (vadimberezniker) [reply server/util/db/db.go:155]: Settings is a sync.Map so the second value indicates whether the key was present in the map.
I'm relying on checking whether v is nil to do the same. I've added a comment to clarify.

File: server/util/db/db.go (hunk 5 above)
Comment (vadimberezniker) [reply server/util/db/db.go]: l is initialized immediately after l is defined so it should never be nil.

File: server/util/db/db.go (hunk 6 above)
Comment (vadimberezniker) [reply server/util/db/db.go]: Done.

**Review by tylerwilliams — APPROVED** [review]
LGTM

**Review by tylerwilliams — APPROVED** [review]
LGTM
